File: competitive_programing/leetcode/problem_100/Sol.java
class Solution {
  public boolean isSameTree(TreeNode p, TreeNode q) {
    List<Integer> l1 = new ArrayList<>(),
        l2 = new ArrayList<>();
    inOrder(p, l1);
    inOrder(q, l2);

    if (l1.size() != l2.size())
      return false;

    for (int i = 0; i < l1.size(); i++)
      if (l1.get(i) != l2.get(i)) {
        if (l1.get(i) != null && l2.get(i) != null && l1.get(i) - l2.get(i) == 0)
          continue;
        return false;
      }

    l1 = null;
    l2 = null;

    return true;
  }

  public void inOrder(TreeNode n, List<Integer> l) {
    l.add(n == null ? null : n.val);
    if (n == null)
      return;
    inOrder(n.left, l);
    inOrder(n.right, l);
  }
}

File: competitive_programing/leetcode/problem_112/sol.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode* root, int targetSum) {
        bool result = false; 
        int currentSum = 0;

        s(root, targetSum, currentSum, result);
        return result;
    }

    void s(TreeNode * n, int& targetSum , int& currentSum, bool& result) {
        if (n == nullptr || result) return;
        currentSum += n->val;

        if ( (n->left == nullptr) && (n->right == nullptr) && currentSum == targetSum) {
            result = true;
            return;
        }
        s(n->left, targetSum, currentSum, result);
        s(n->right, targetSum, currentSum, result);
        currentSum -= n->val;
    }
};
